### PR TITLE
adding dns support for dnspod provider

### PR DESCRIPTION
**Pull request by piratimir**
### Description

I coded the dns driver for DNSPod on libcloud/dns/drivers/dnspod.py module. In order for this driver to work the libcloud/common/dnspod.py module is needed. The unitttests for this driver are coded in the libcloud/test/dns/test_dnspod.py module.
### Status
- done, ready for review
### Checklist (tick everything that applies)
- [ ] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [ ] Documentation
- [ ] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [ ] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

### CODE REVIEW

**Comment by Kami** [comment]
Great, thanks.

Is the PR already for a review, or are you still working on it?

**Comment by piratimir** [comment]
@Kami the pr is ready for review :)

File: libcloud/test/dns/test_dnspod.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    def _Record_Remove_DELETE_RECORD_SUCCESS(self, method, url, body, headers):
+        body = self.fixtures.load('delete_record_success.json')
+
+        return httplib.OK, body, {}, httplib.responses[httplib.OK]
+
+    def _Record_Remove_DELETE_RECORD_RECORD_DOES_NOT_EXIST(self, method,
+                                                           url, body, headers):
+        body = self.fixtures.load('delete_record_record_does_not_exist.json')
+
+        return httplib.OK, body, {}, httplib.responses[httplib.OK]
```
Comment (Kami) [inline libcloud/test/dns/test_dnspod.py:240]: Just curious - does API indeed return 200 status code for all the different responses?

File: libcloud/common/dnspod.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+from libcloud.common.base import ConnectionKey, JsonResponse
```
Comment (Kami) [inline libcloud/common/dnspod.py:16]: Please add a missing Apache license header.

File: libcloud/common/dnspod.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class DNSPodConnection(ConnectionKey):
+    host = API_HOST
+    responseCls = DNSPodResponse
+
+    def add_default_headers(self, headers):
+        headers['Content-Type'] = 'application/x-www-form-urlencoded'
+        headers['Accept'] = 'text/json'
+        headers['User-Agent'] = \
```
Comment (Kami) [inline libcloud/common/dnspod.py]: Please remove this, libcloud includes it's own custom User-Agent header (it includes version of Libcloud and driver user).

File: libcloud/dns/drivers/dnspod.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+        RecordType.NS: 'NS',
+        RecordType.PTR: 'PTR',
+        RecordType.SOA: 'SOA',
+        RecordType.SRV: 'SRV',
+        RecordType.TXT: 'TXT'
+    }
+
+    def list_zones(self):
+        action = '/Domain.List'
+        data = {'user_token': self.key, 'format': 'json'}
```
Comment (Kami) [inline libcloud/dns/drivers/dnspod.py]: `user_token` and `format` attributes seem to be included in every request so you could maybe refactor this a bit and include those attributes somewhere else (e.g. inside the connection `request` method or similar) so the same piece of code doesn't need to be repeated all over.

File: libcloud/dns/drivers/dnspod.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+        data = urlencode(data)
+        try:
+            response = self.connection.request(action=action, method='POST',
+                                               data=data)
+        except DNSPodException:
+            e = sys.exc_info()[1]
+            if e.message == 'No domains':
+                return []
+        zones = self._to_zones(items=response.parse_body()['domains'])
```
Comment (Kami) [inline libcloud/dns/drivers/dnspod.py]: Libcloud already parses the body for you and is available via `response.object` attribute.

No need to call `parse_body` yourself.

File: libcloud/dns/drivers/dnspod.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+        data = {'user_token': self.key, 'format': 'json', 'domain': domain}
+        if extra is not None:
+            data.update(extra)
+        data = urlencode(data)
+        try:
+            response = self.connection.request(action=action, method='POST',
+                                               data=data)
+        except DNSPodException:
+            e = sys.exc_info()[1]
+            if e.message in ['Domain is exists',
```
Comment (Kami) [inline libcloud/dns/drivers/dnspod.py]: You could save those messages in a constant, e.g. `ZONE_ALREADY_EXISTS_ERROR_MSGS = [ ... ]`.

**Comment by Kami** [comment]
Thanks :+1: 

I've added some in-line comments.

File: libcloud/dns/drivers/dnspod.py (hunk 4 above)
Comment (piratimir) [inline libcloud/dns/drivers/dnspod.py]: @Kami yes i thought about that, how to include that in every request?

Comment (piratimir) [inline libcloud/dns/drivers/dnspod.py]: @Kami hey man any update on this? I have fixed the other problems, this is the only one left.

Comment (Kami) [inline libcloud/dns/drivers/dnspod.py]: As mentioned in my comment above, you could refactor that functionality in a custom `request` method on the connection class.

You can find a similar example here - https://github.com/apache/libcloud/blob/trunk/libcloud/dns/drivers/cloudflare.py#L106

**Comment by Kami** [comment]
@jetbird Did you push your latest changes?

Since I don't see any new commits (license header is still missing, etc.).

**Comment by piratimir** [comment]
@Kami  done :)

**Comment by Kami** [comment]
@jetbird Thanks.

I've made some changes and fixes (6594a46c53ef6d9953b742826c4fc5d86bc0ed58, defff8acd1028f78b58df32e8dcd179f728aaf0c) and merged patch into trunk.

As you can see in the second commit (defff8acd1028f78b58df32e8dcd179f728aaf0c), `_make_request` method didn't actually do anything with the `method` argument (a bug which existing tests don't catch). That was masked by (ab)using `**kwargs` - that's a good example of why `**kwargs` should be avoided in scenarios like that and only used when there is a valid need for it.

**Comment by Kami** [comment]
@jetbird Can you please test latest changes against the live provider API?

**Comment by piratimir** [comment]
Thanks for explaining the fixes :) Of course, I will test it now.

**Comment by piratimir** [comment]
@Kami Tested. It works fine for me :)

**Comment by Kami** [comment]
@jetbird Great :+1: 

It would also be good if you can add some basic docs for the driver and contact the provider (tall them about Libcloud, driver, etc. and ask them if it's possible to mention it somewhere on the page / docs, etc.).

**Pull request closed without merging**